ui/src/app/utils/cluster-initializer.spec.tsx

```diff
@@ -0,0 +1,305 @@
+      // cluster statuses mock out what we might expect to see after that call.
+      {...baseCluster, status: ClusterStatus.Starting},
+      {...baseCluster, status: ClusterStatus.Running}
+    ]);
+    const initializer = createInitializer();
+    const cluster = await runInitializerAndTimers(initializer);
+    expect(mockStartCluster).toHaveBeenCalled();
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('should call callback with cluster status', async() => {
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Stopped},
+      {...baseCluster, status: ClusterStatus.Starting},
+      {...baseCluster, status: ClusterStatus.Running}
+    ]);
+    const mockFn = jest.fn();
+    const initializer = createInitializer({
+      onStatusUpdate: mockFn
+    });
+
+    const cluster = await runInitializerAndTimers(initializer);
+    expect(mockFn.mock.calls.length).toEqual(3);
+    expect(mockFn.mock.calls).toEqual([
+      [ClusterStatus.Stopped],
+      [ClusterStatus.Starting],
+      [ClusterStatus.Running]]
+    );
+  });
+
+  it('should create cluster if it is initially nonexistent', async() => {
+    mockGetCluster.mockImplementationOnce((workspaceNamespace) => {
+      throw new Response(null, {status: 404});
+    });
+    mockCreateCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      return {...baseCluster, status: ClusterStatus.Creating};
+    });
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Starting},
+      {...baseCluster, status: ClusterStatus.Running}
+    ]);
+    const initializer = createInitializer();
+    const cluster = await runInitializerAndTimers(initializer);
+
+    expect(mockCreateCluster).toHaveBeenCalled();
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('should delete cluster if in an error state', async() => {
+    // A cluster in an error state should trigger a deletion request.
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Creating},
+      {...baseCluster, status: ClusterStatus.Error},
+    ]);
+    mockDeleteCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      return {};
+    });
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Deleting},
+      {...baseCluster, status: ClusterStatus.Deleting},
+    ]);
+    // After some period of "deleting" status, we expect the cluster to become nonexistent...
+    mockGetCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      throw new Response(null, {status: 404});
+    });
+    // which should trigger a creation request...
+    mockCreateCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      return {...baseCluster, status: ClusterStatus.Creating};
+    });
+    // and eventually give us a good cluster.
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Starting},
+      {...baseCluster, status: ClusterStatus.Running}
+    ]);
+
```

Comment:
nit: I wonder if there's a way to visually separate the setup from the punchline in these longer tests. The code under test is almost an afterthought in terms of how little space it takes up.